src/main/kotlin/org/jitsi/jicofo/conference/ColibriSessionManager.kt

```diff
@@ -0,0 +1,366 @@
+        }
+    }
+
+    private fun getBridges(): Map<Bridge, Int> = synchronized(syncRoot) {
+        return bridgeSessions.filter { !it.hasFailed }.associate { Pair(it.bridge, it.participants.size) }
+    }
+
+    private fun selectBridge(participant: Participant): Bridge {
+        if (findBridgeSession(participant) != null) {
+            // This should never happen.
+            throw IllegalStateException("The participant already has a bridge:" + participant.chatMember.name)
+        }
+
+        return jicofoServices.bridgeSelector.selectBridge(getBridges(), participant.chatMember.region)
+            ?: throw BridgeSelectionFailedException()
+    }
+
+    private fun updateOctoRelays() {
+        if (!OctoConfig.config.enabled) {
+            return
+        }
+
+        synchronized(syncRoot) {
+            val allRelays = getAllRelays(null)
+
+            logger.info("Updating Octo relays: $allRelays")
+            operationalBridges().forEach { it.setRelays(allRelays) }
+        }
+    }
+
+    private fun removeNonOperationalBridges() {
+        val nonOperationalBridges: Set<Jid> = bridgeSessions
+            .filter { it.hasFailed || !it.bridge.isOperational }
+            .map { it.bridge.jid }.toSet()
+
+        if (nonOperationalBridges.isNotEmpty()) {
+            bridgesDown(nonOperationalBridges)
+        }
+    }
+
+    private fun findBridgeSession(jid: Jid) = bridgeSessions.find { it.bridge.jid == jid }
+    private fun findBridgeSession(bridge: Bridge) = bridgeSessions.find { it.bridge == bridge }
+
+    /**
+     * Update octo channels on all bridges except `except`, removing the specified set of `sources`.
+     * @param sources the sources to remove.
+     * @param except the bridge session which is not to be updated.
+     */
+    private fun removeSourcesFromOcto(sources: ConferenceSourceMap, except: BridgeSession) = synchronized(syncRoot) {
+        operationalBridges()
+            .filter { it != except }
+            .forEach { it.removeSourcesFromOcto(sources) }
+    }
+
+    /**
+     * Adds the specified sources and source groups to the Octo participants
+     * of all bridges except for `exclude`.
+     * @param except the bridge to which sources will not be added (i.e. the
+     * bridge to which the participant whose sources we are adding is
+     * connected).
+     * @param sources the sources to add.
+     */
+    private fun addSourcesToOcto(sources: ConferenceSourceMap, except: BridgeSession) = synchronized(sources) {
+        operationalBridges()
+            .filter { it != except }
+            .forEach { it.addSourcesToOcto(sources) }
+    }
+
+    /**
+     * Get a stream of those bridges which are operational.
+     */
+    private fun operationalBridges(): List<BridgeSession> = synchronized(syncRoot) {
+        bridgeSessions.filter { !it.hasFailed && it.bridge.isOperational }
+    }
+
+    /**
+     * Interface for events fired by [ColibriSessionManager].
+     */
```

Comment:
Maybe comment that the `ColibriSessionManager` calls these while holding its internal mutexes so listeners should be careful not to perform actions that'll cause deadlocks?